synapse/lib/boss.py

```diff
@@ -18,8 +21,40 @@ class Boss(s_base.Base):
     async def __anit__(self):
         await s_base.Base.__anit__(self)
         self.tasks = {}
+        self.shutting = False
         self.onfini(self._onBossFini)
 
+    async def shutdown(self, timeout=None):
+        # when a boss is "shutting down" it should not promote any new tasks,
+        # but await the completion of any which are already underway...
+        self.shutting = True
+
+        for task in list(self.tasks.values()):
+
+            # do not wait on child tasks
+            if task.root is not None:
+                continue
+
+            # do not wait on background tasks
+            if task.background:
+                continue
+
+            if not await s_coro.waittask(task.task, timeout=timeout):
+                self.shutting = False
```

Comment:
This condition means we're unable to shutdown - how would we communicate that failure to the operator?